llvm/test/CodeGen/RISCV/addimm-mulimm.ll

```diff
@@ -49,19 +53,24 @@ define i64 @add_mul_combine_accept_a3(i64 %x) {
 ; RV32IMB-LABEL: add_mul_combine_accept_a3:
 ; RV32IMB:       # %bb.0:
 ; RV32IMB-NEXT:    li a2, 29
-; RV32IMB-NEXT:    mul a1, a1, a2
-; RV32IMB-NEXT:    mulhu a3, a0, a2
-; RV32IMB-NEXT:    add a1, a3, a1
-; RV32IMB-NEXT:    mul a2, a0, a2
+; RV32IMB-NEXT:    mulhu a2, a0, a2
```

Comment:
We may have latency=2 multiply instructions on a high performance RV64 core, so this may be an regression. But this is a RV32 test (multiply won't be so fast), so I just leave this comment for record.

Reply:
FYI, Syntacore SCR1 has single-cycle integer multiplier :(

Reply:
A couple things here.

First, I agree that we are likely to want a per processor cost based threshold here.  The only question for me is order of work and relative importance, not net result.

I wasn't aware SCR1 has a single cycle latency.  I glanced at the schedule model for that just now, and given the only latencies set appear to be div/rem, I have to ask.  Are you sure this is actually true?

Second, I'll argue that instruction count != latency.  In this particular case, we have two independent instructions followed by a dependent subtract.  On any core with multiple issue and single cycle shl/add, I'd expect that to have a latency of 2.  Obviously, this argument can be taken to a ridiculous extreme, but it isn't obvious to me that a core with single cycle mul only wants single cycle expansions.  The SCR1 might be a weird example here - in the schedule model it looks like there's only one ALU?  Even then, that's not solely a function of latency.  

For comparison, gcc appears to be defaulting to an expansion budget of at least four instructions for rv64gc.

Reply:
I'm not opposed to this optimization.
For high performance core (the latency will be 3-4 normally, 2 is too extreme), the units for simple ALU should be more than multiplier units, so this optimization should be an uplift.
For RV32 core, they are normally single issue with fewer units and the latency of multiplier can be very small because of low frequency (if you agree), I think this optimization will be a decrease.

> First, I agree that we are likely to want a per processor cost based threshold here.
> For comparison, gcc appears to be defaulting to an expansion budget of at least four instructions for rv64gc. (example: https://godbolt.org/z/zzvqfha8f)

Yeah, I agree too. Actually this is what GCC is doing. The cost of mul is 2 for generic OoO, 1 for size, 4 for rocket and others (https://godbolt.org/z/8nvdbo5s1)
https://github.com/gcc-mirror/gcc/blob/58a0b190a256bd2a184554de0fae0031a614ec67/gcc/config/riscv/riscv.cc#L275-L292
We may add similar subtarget feature in LLVM too.